rfcs/2021-08-13-8025-internal-tracing.md

```diff
@@ -0,0 +1,254 @@
+# RFC #8025 - 2021-08-12 - Internal Tracing
+
+This RFC discusses an integration with the Datadog APM product for Vector's
+internal tracing data, with the goal of enabling faster and more accurate
+diagnosis of performance issues in deployed environments.
+
+## Context
+
+- Existing experimental implementation ([#7929](https://github.com/timberio/vector/pull/7929))
+- Internal instrumentation epic ([#3900](https://github.com/timberio/vector/issues/3900))
+
+## Scope
+
+### In scope
+
+- Instrumenting components with tracing spans
+- Sending internal tracing spans to Datadog APM via the Datadog Agent
+- An opt-in system for internal soak tests and experimental use in customer environments
+
+### Out of scope
+
+- Sending internal traces anywhere other than Datadog APM
+- Sending internal traces directly to Datadog without a local agent
+- An always-on tracing subsystem that is enabled by default in all customer deployments
+- General support for sending tracing data through Vector (i.e. trace data type, sources and sinks)
+
+## Pain
+
+- Debugging Vector performance in users environments requires a lot of back and forth
+- The data available for debugging Vector is relatively high level and doesn't always map clearly to specific components and operations
+
+## Proposal
+
+### User Experience
+
+Vector will gain a new `--enable-datadog-tracing` flag that users can enable at
+start time. When enabled, Vector will forward internal tracing spans to a local
+Datadog Agent with the APM feature enabled. A valid API key will be required to
+start the agent, but no additional configuration of Vector itself is required.
+
+### Implementation
+
+#### Overview
+
+Vector already has some very basic trace-style instrumentation via the
+[`tracing`](https://docs.rs/tracing/) crate that we currently use primary for
+logging. This RFC proposes that we expand that instrumentation to include more
+granular spans around the existing events (i.e. `info!` and friends, not Vector
+events).
```

Comment:
Vector events call `info!`, correct? In other words, all of Vector's event-driven log lines will be included?

Reply:
There are three things called events here:

1. Vector's `Event`: the actual user data flowing through the system
2. Vector's `internal_events`: the abstraction we built for emitting internal observability data
3. The `tracing` crate's `Event`: the data created with `info!` and friends, that manifest as Vector's own logging

We're not talking about shipping any of these as part of tracing, just the spans that surround and give context to (3).